File: src/bin/19_demo.cc
// 19_demo.cc: C++17 版本的学生管理 CLI（CRUD + search + order）。
//
// 编译：
//   g++ -std=c++17 -O2 -Wall -Wextra -pedantic src/bin/19_demo.cc -o
//   /tmp/19_demo_cc
//
// 运行：
//   /tmp/19_demo_cc
//
// 命令：
// - add <name> <age> <class>
// - list
// - remove <id>
// - mod <id> <name> <age> <class>
// - search id <id>
// - search name <name>
// - order <id|name|age|class> <asc|desc>
// - help
// - quit / exit

#include <algorithm>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <map>
#include <set>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

struct Student {
  std::uint32_t id;
  std::string name;
  std::uint8_t age;
  std::string class_name;
};

// 存储设计：
// - 主数据放在 by_id（unordered_map）里。
// - ids / name_index 是轻量索引，仅存 id，便于有序遍历和按 name 查询。
class StudentStore {
 public:
  std::uint32_t Add(const std::string& name, std::uint8_t age,
                    const std::string& class_name) {
    const std::uint32_t id = next_id_++;
    Student s{id, name, age, class_name};
    by_id_.emplace(id, s);
    ids_.insert(id);
    name_index_[name].insert(id);
    return id;
  }

  const Student* GetById(std::uint32_t id) const {
    auto it = by_id_.find(id);
    if (it == by_id_.end()) return nullptr;
    return &it->second;
  }

  std::vector<const Student*> ListById() const {
    std::vector<const Student*> out;
    out.reserve(ids_.size());
    for (std::uint32_t id : ids_) {
      auto it = by_id_.find(id);
      if (it != by_id_.end()) out.push_back(&it->second);
    }
    return out;
  }

  std::vector<const Student*> SearchByNameExact(const std::string& name) const {
    std::vector<const Student*> out;
    auto it = name_index_.find(name);
    if (it == name_index_.end()) return out;
    out.reserve(it->second.size());
    for (std::uint32_t id : it->second) {
      auto hit = by_id_.find(id);
      if (hit != by_id_.end()) out.push_back(&hit->second);
    }
    return out;
  }

  bool Remove(std::uint32_t id) {
    auto it = by_id_.find(id);
    if (it == by_id_.end()) return false;

    RemoveNameIndex(it->second.name, id);
    ids_.erase(id);
    by_id_.erase(it);
    return true;
  }

  bool Modify(std::uint32_t id, const std::string& name, std::uint8_t age,
              const std::string& class_name) {
    auto it = by_id_.find(id);
    if (it == by_id_.end()) return false;

    std::string old_name = it->second.name;
    if (old_name != name) {
      RemoveNameIndex(old_name, id);
      name_index_[name].insert(id);
    }

    it->second.name = name;
    it->second.age = age;
    it->second.class_name = class_name;
    return true;
  }

  std::vector<const Student*> Ordered(const std::string& field,
                                      const std::string& direction) const {
    std::vector<const Student*> rows;
    rows.reserve(by_id_.size());
    for (const auto& kv : by_id_) rows.push_back(&kv.second);

    auto cmp_asc = [&](const Student* a, const Student* b) {
      if (field == "id") return a->id < b->id;
      if (field == "name")
        return (a->name == b->name) ? (a->id < b->id) : (a->name < b->name);
      if (field == "age")
        return (a->age == b->age) ? (a->id < b->id) : (a->age < b->age);
      // class
      return (a->class_name == b->class_name) ? (a->id < b->id)
                                              : (a->class_name < b->class_name);
    };

    std::sort(rows.begin(), rows.end(), cmp_asc);
    if (direction == "desc") std::reverse(rows.begin(), rows.end());
    return rows;
  }

 private:
  void RemoveNameIndex(const std::string& name, std::uint32_t id) {
    auto it = name_index_.find(name);
    if (it == name_index_.end()) return;
    it->second.erase(id);
    if (it->second.empty()) name_index_.erase(it);
  }

  std::unordered_map<std::uint32_t, Student> by_id_;
  std::set<std::uint32_t> ids_;
  std::map<std::string, std::set<std::uint32_t>> name_index_;
  std::uint32_t next_id_ = 1;
};

static void PrintHelp() {
  std::cout << "commands:\n";
  std::cout << "  add <name> <age> <class>              add a student\n";
  std::cout
      << "  list                                  list all students by id\n";
  std::cout << "  remove <id>                           remove by id\n";
  std::cout << "  mod <id> <name> <age> <class>         modify by id\n";
  std::cout << "  search id <id>                        search by id\n";
  std::cout << "  search name <name>                    search by exact name\n";
  std::cout << "  order <id|name|age|class> <asc|desc>  ordered view\n";
  std::cout << "  help                                  show help\n";
  std::cout << "  quit | exit                           leave repl\n";
}

static void PrintStudents(const std::vector<const Student*>& rows) {
  if (rows.empty()) {
    std::cout << "(empty)\n";
    return;
  }
  std::cout << std::left << std::setw(4) << "id" << " " << std::setw(12)
            << "name" << " " << std::setw(4) << "age" << " " << std::setw(12)
            << "class" << "\n";
  for (const Student* s : rows) {
    std::cout << std::left << std::setw(4) << s->id << " " << std::setw(12)
              << s->name << " " << std::setw(4) << static_cast<int>(s->age)
              << " " << std::setw(12) << s->class_name << "\n";
  }
}

static bool ParseUint32(const std::string& raw, const char* label,
                        std::uint32_t* out) {
  std::istringstream iss(raw);
  std::uint64_t v = 0;
  if (!(iss >> v) || !iss.eof() || v > UINT32_MAX) {
    std::cout << "error: invalid " << label << " `" << raw << "`\n";
    return false;
  }
  *out = static_cast<std::uint32_t>(v);
  return true;
}

static bool ParseAge(const std::string& raw, std::uint8_t* out) {
  std::uint32_t v = 0;
  if (!ParseUint32(raw, "age", &v) || v > UINT8_MAX) {
    std::cout << "error: invalid age `" << raw << "`\n";
    return false;
  }
  *out = static_cast<std::uint8_t>(v);
  return true;
}

static std::vector<std::string> Split(const std::string& line) {
  std::istringstream iss(line);
  std::vector<std::string> parts;
  std::string token;
  while (iss >> token) parts.push_back(token);
  return parts;
}

// 返回 true 表示继续循环，false 表示退出。
static bool HandleCommand(const std::string& line, StudentStore* store) {
  std::vector<std::string> parts = Split(line);
  if (parts.empty()) return true;

  const std::string& cmd = parts[0];
  if (cmd == "add") {
    if (parts.size() != 4) {
      std::cout << "usage: add <name> <age> <class>\n";
      return true;
    }
    std::uint8_t age = 0;
    if (!ParseAge(parts[2], &age)) return true;
    std::uint32_t id = store->Add(parts[1], age, parts[3]);
    std::cout << "ok: added id=" << id << "\n";
    return true;
  }

  if (cmd == "list") {
    if (parts.size() != 1) {
      std::cout << "usage: list\n";
      return true;
    }
    PrintStudents(store->ListById());
    return true;
  }

  if (cmd == "remove") {
    if (parts.size() != 2) {
      std::cout << "usage: remove <id>\n";
      return true;
    }
    std::uint32_t id = 0;
    if (!ParseUint32(parts[1], "id", &id)) return true;
    if (store->Remove(id)) {
      std::cout << "ok: removed id=" << id << "\n";
    } else {
      std::cout << "error: id=" << id << " not found\n";
    }
    return true;
  }

  if (cmd == "mod") {
    if (parts.size() != 5) {
      std::cout << "usage: mod <id> <name> <age> <class>\n";
      return true;
    }
    std::uint32_t id = 0;
    std::uint8_t age = 0;
    if (!ParseUint32(parts[1], "id", &id)) return true;
    if (!ParseAge(parts[3], &age)) return true;
    if (store->Modify(id, parts[2], age, parts[4])) {
      std::cout << "ok: modified id=" << id << "\n";
    } else {
      std::cout << "error: id=" << id << " not found\n";
    }
    return true;
  }

  if (cmd == "search") {
    if (parts.size() < 3) {
      std::cout << "usage: search id <id> | search name <name>\n";
      return true;
    }
    if (parts[1] == "id") {
      if (parts.size() != 3) {
        std::cout << "usage: search id <id>\n";
        return true;
      }
      std::uint32_t id = 0;
      if (!ParseUint32(parts[2], "id", &id)) return true;
      const Student* s = store->GetById(id);
      if (s == nullptr) {
        std::cout << "(empty)\n";
      } else {
        PrintStudents({s});
      }
      return true;
    }
    if (parts[1] == "name") {
      if (parts.size() != 3) {
        std::cout << "usage: search name <name>\n";
        return true;
      }
      PrintStudents(store->SearchByNameExact(parts[2]));
      return true;
    }
    std::cout << "usage: search id <id> | search name <name>\n";
    return true;
  }

  if (cmd == "order") {
    if (parts.size() != 3) {
      std::cout << "usage: order <id|name|age|class> <asc|desc>\n";
      return true;
    }
    if (parts[1] != "id" && parts[1] != "name" && parts[1] != "age" &&
        parts[1] != "class") {
      std::cout << "error: invalid field `" << parts[1] << "`\n";
      return true;
    }
    if (parts[2] != "asc" && parts[2] != "desc") {
      std::cout << "error: invalid direction `" << parts[2] << "`\n";
      return true;
    }
    PrintStudents(store->Ordered(parts[1], parts[2]));
    return true;
  }

  if (cmd == "help") {
    PrintHelp();
    return true;
  }

  if (cmd == "quit" || cmd == "exit") {
    return false;
  }

  std::cout << "unknown command. type `help`\n";
  return true;
}

int main() {
  StudentStore store;

  std::cout << "student-cli demo (cpp17)\n";
  std::cout << "type `help` to see commands\n";

  std::string line;
  while (true) {
    std::cout << "sms> " << std::flush;
    if (!std::getline(std::cin, line)) {
      std::cout << "\n";
      break;
    }
    if (line.empty()) continue;
    if (!HandleCommand(line, &store)) break;
  }

  std::cout << "bye\n";
  return 0;
}

File: src/bin/19_demo.rs
//! 19_demo: 通过一个 CLI 小练习串起基础业务逻辑（CRUD + 查询 + 排序）。
//!
//! 运行：
//! cargo run --bin 19_demo
//!
//! 命令：
//! - add <name> <age> <class>
//! - list
//! - remove <id>
//! - mod <id> <name> <age> <class>
//! - search id <id>
//! - search name <name>
//! - order <id|name|age|class> <asc|desc>
//! - help
//! - quit / exit

use std::cmp::Ordering;
use std::collections::{BTreeMap, BTreeSet, HashMap};
use std::io::{self, Write};

#[derive(Debug, Clone)]
struct Student {
    id: u32,
    name: String,
    age: u8,
    class_name: String,
}

#[derive(Debug, Default)]
// 存储设计总览：
// - 主数据只放在 `by_id`（HashMap）里，这是唯一完整记录存储。
// - `ids`/`name_index` 是轻量索引层，只保存 id（主键）来加速查询/排序。
// - 可以理解为“一份主存 + 多份索引”，而不是复制多份 Student 全量数据。
// - 思路上有点像 arena 的“句柄化访问”（用 id 回主存取值），
//   但这里本质是索引化存储，不是 arena allocator。
struct StudentStore {
    // 主索引：按 id O(1) 查找。
    by_id: HashMap<u32, Student>,
    // 有序 id 索引：便于稳定 list（默认按 id 升序）。
    ids: BTreeSet<u32>,
    // 名称索引：便于 `search name <name>` 精确匹配。
    name_index: BTreeMap<String, BTreeSet<u32>>,
    next_id: u32,
}

impl StudentStore {
    fn new() -> Self {
        Self {
            by_id: HashMap::new(),
            ids: BTreeSet::new(),
            name_index: BTreeMap::new(),
            next_id: 1,
        }
    }

    fn add(&mut self, name: &str, age: u8, class_name: &str) -> u32 {
        let id = self.next_id;
        self.next_id += 1;

        let student = Student {
            id,
            name: name.to_string(),
            age,
            class_name: class_name.to_string(),
        };

        self.by_id.insert(id, student.clone());
        self.ids.insert(id);
        self.name_index
            .entry(student.name.clone())
            .or_default()
            .insert(id);

        id
    }

    fn get_by_id(&self, id: u32) -> Option<&Student> {
        self.by_id.get(&id)
    }

    fn list_by_id(&self) -> Vec<&Student> {
        self.ids
            .iter()
            .filter_map(|id| self.by_id.get(id))
            .collect::<Vec<&Student>>()
    }

    fn search_by_name_exact(&self, name: &str) -> Vec<&Student> {
        match self.name_index.get(name) {
            Some(id_set) => id_set
                .iter()
                .filter_map(|id| self.by_id.get(id))
                .collect::<Vec<&Student>>(),
            None => Vec::new(),
        }
    }

    fn remove(&mut self, id: u32) -> bool {
        let removed = match self.by_id.remove(&id) {
            Some(v) => v,
            None => return false,
        };

        self.ids.remove(&id);
        self.remove_name_index(&removed.name, id);
        true
    }

    fn modify(&mut self, id: u32, name: &str, age: u8, class_name: &str) -> bool {
        let old_name = match self.by_id.get(&id) {
            Some(v) => v.name.clone(),
            None => return false,
        };
        if old_name != name {
            self.remove_name_index(&old_name, id);
            self.name_index
                .entry(name.to_string())
                .or_default()
                .insert(id);
        }

        let student = self
            .by_id
            .get_mut(&id)
            .expect("id checked above, get_mut must succeed");
        student.name = name.to_string();
        student.age = age;
        student.class_name = class_name.to_string();
        true
    }

    fn remove_name_index(&mut self, name: &str, id: u32) {
        if let Some(set) = self.name_index.get_mut(name) {
            set.remove(&id);
            if set.is_empty() {
                self.name_index.remove(name);
            }
        }
    }
}

#[derive(Debug, Clone, Copy)]
enum SortField {
    Id,
    Name,
    Age,
    Class,
}

#[derive(Debug, Clone, Copy)]
enum SortDirection {
    Asc,
    Desc,
}

fn print_help() {
    println!("commands:");
    println!("  add <name> <age> <class>              add a student");
    println!("  list                                  list all students by id");
    println!("  remove <id>                           remove by id");
    println!("  mod <id> <name> <age> <class>         modify by id");
    println!("  search id <id>                        search by id (O(1) index)");
    println!("  search name <name>                    search by exact name");
    println!("  order <id|name|age|class> <asc|desc>  ordered view");
    println!("  help                                  show help");
    println!("  quit | exit                           leave repl");
}

fn print_students(students: &[&Student]) {
    if students.is_empty() {
        println!("(empty)");
        return;
    }

    println!("{:<4} {:<12} {:<4} {:<12}", "id", "name", "age", "class");
    for s in students {
        println!(
            "{:<4} {:<12} {:<4} {:<12}",
            s.id, s.name, s.age, s.class_name
        );
    }
}

fn parse_id(raw: &str) -> Option<u32> {
    match raw.parse::<u32>() {
        Ok(v) => Some(v),
        Err(_) => {
            println!("error: invalid id `{raw}`");
            None
        }
    }
}

fn parse_age(raw: &str) -> Option<u8> {
    match raw.parse::<u8>() {
        Ok(v) => Some(v),
        Err(_) => {
            println!("error: invalid age `{raw}`");
            None
        }
    }
}

fn parse_sort_field(raw: &str) -> Option<SortField> {
    match raw {
        "id" => Some(SortField::Id),
        "name" => Some(SortField::Name),
        "age" => Some(SortField::Age),
        "class" => Some(SortField::Class),
        _ => {
            println!("error: invalid field `{raw}`");
            None
        }
    }
}

fn parse_sort_direction(raw: &str) -> Option<SortDirection> {
    match raw {
        "asc" => Some(SortDirection::Asc),
        "desc" => Some(SortDirection::Desc),
        _ => {
            println!("error: invalid direction `{raw}`");
            None
        }
    }
}

fn order_students(store: &StudentStore, field: SortField, direction: SortDirection) {
    let mut rows = store.by_id.values().collect::<Vec<&Student>>();
    rows.sort_by(|a, b| {
        let ord = match field {
            SortField::Id => a.id.cmp(&b.id),
            SortField::Name => a.name.cmp(&b.name).then_with(|| a.id.cmp(&b.id)),
            SortField::Age => a.age.cmp(&b.age).then_with(|| a.id.cmp(&b.id)),
            SortField::Class => a
                .class_name
                .cmp(&b.class_name)
                .then_with(|| a.id.cmp(&b.id)),
        };
        match direction {
            SortDirection::Asc => ord,
            SortDirection::Desc => match ord {
                Ordering::Less => Ordering::Greater,
                Ordering::Equal => Ordering::Equal,
                Ordering::Greater => Ordering::Less,
            },
        }
    });
    print_students(&rows);
}

// 返回值：true 表示继续循环；false 表示退出。
fn handle_command(line: &str, store: &mut StudentStore) -> bool {
    let parts = line.split_whitespace().collect::<Vec<&str>>();
    if parts.is_empty() {
        return true;
    }

    match parts[0] {
        "add" => {
            if parts.len() != 4 {
                println!("usage: add <name> <age> <class>");
                return true;
            }
            if let Some(age) = parse_age(parts[2]) {
                let id = store.add(parts[1], age, parts[3]);
                println!("ok: added id={id}");
            }
        }
        "list" => {
            if parts.len() != 1 {
                println!("usage: list");
                return true;
            }
            let rows = store.list_by_id();
            print_students(&rows);
        }
        "remove" => {
            if parts.len() != 2 {
                println!("usage: remove <id>");
                return true;
            }
            if let Some(id) = parse_id(parts[1]) {
                if store.remove(id) {
                    println!("ok: removed id={id}");
                } else {
                    println!("error: id={id} not found");
                }
            }
        }
        // 这里命令名写 `mod`，仅是字符串命令，和 Rust 关键字不冲突。
        "mod" => {
            if parts.len() != 5 {
                println!("usage: mod <id> <name> <age> <class>");
                return true;
            }
            let id = match parse_id(parts[1]) {
                Some(v) => v,
                None => return true,
            };
            let age = match parse_age(parts[3]) {
                Some(v) => v,
                None => return true,
            };
            if store.modify(id, parts[2], age, parts[4]) {
                println!("ok: modified id={id}");
            } else {
                println!("error: id={id} not found");
            }
        }
        "search" => {
            if parts.len() < 3 {
                println!("usage: search id <id> | search name <name>");
                return true;
            }
            match parts[1] {
                "id" => {
                    if parts.len() != 3 {
                        println!("usage: search id <id>");
                        return true;
                    }
                    let id = match parse_id(parts[2]) {
                        Some(v) => v,
                        None => return true,
                    };
                    match store.get_by_id(id) {
                        Some(s) => print_students(&[s]),
                        None => println!("(empty)"),
                    }
                }
                "name" => {
                    if parts.len() != 3 {
                        println!("usage: search name <name>");
                        return true;
                    }
                    let rows = store.search_by_name_exact(parts[2]);
                    print_students(&rows);
                }
                _ => println!("usage: search id <id> | search name <name>"),
            }
        }
        "order" => {
            if parts.len() != 3 {
                println!("usage: order <id|name|age|class> <asc|desc>");
                return true;
            }
            let field = match parse_sort_field(parts[1]) {
                Some(v) => v,
                None => return true,
            };
            let direction = match parse_sort_direction(parts[2]) {
                Some(v) => v,
                None => return true,
            };
            order_students(store, field, direction);
        }
        "help" => print_help(),
        "quit" | "exit" => return false,
        _ => println!("unknown command. type `help`"),
    }

    true
}

fn main() -> io::Result<()> {
    // 内存态 demo：不落盘，退出后数据清空。
    let mut store = StudentStore::new();

    println!("student-cli demo");
    println!("type `help` to see commands");

    let stdin = io::stdin();
    loop {
        print!("sms> ");
        io::stdout().flush()?;

        let mut line = String::new();
        if stdin.read_line(&mut line)? == 0 {
            // EOF（如 Ctrl-D）时退出。
            println!();
            break;
        }
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        if !handle_command(line, &mut store) {
            break;
        }
    }

    println!("bye");
    Ok(())
}
